clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -709,12 +711,25 @@ private boolean appendsInProgress() {
     }
 
     /**
-     * Mark all partitions as ready to send and block until the send is complete
+     * Mark all partitions as ready to send and block until the send is complete or time expires
      */
-    public void awaitFlushCompletion() throws InterruptedException {
+    public void awaitFlushCompletion(long timeoutMs) throws InterruptedException {
         try {
-            for (ProducerBatch batch : this.incomplete.copyAll())
-                batch.produceFuture.await();
+            Long expireMs = System.currentTimeMillis() + timeoutMs;
+            int numBatchesFlushed = 0;
+            for (ProducerBatch batch : this.incomplete.copyAll()) {
+                Long currentMs = System.currentTimeMillis();
+                if (currentMs > expireMs) {
+                    throw new TimeoutException(String.format("Failed to flush accumulated records within %d milliseconds,"
+                        + " successfully completed %d batches.", timeoutMs, numBatchesFlushed));
+                }
+                boolean completed = batch.produceFuture.await(Math.max(expireMs - currentMs, 0), TimeUnit.MILLISECONDS);
+                if (!completed) {
+                    throw new TimeoutException(String.format("Failed to flush accumulated records within %d milliseconds,"
```

Comment:
`Failed to flush accumulated records` -> `Failed to flush all accumulated records...`

Also, it would be much more useful if we could say `%d of %d` batches although I see we would have to expose a `size()` method in `IncompleteBatches`